src/common.h

```diff
@@ -6,12 +6,83 @@
 
 #pragma once
 
+#include <stdarg.h>
+#include <string.h>
 #include "redismodule.h"
 #if defined(DEBUG) || !defined(NDEBUG)
 #include "readies/cetara/diag/gdb.h"
 #endif
 
+#ifdef REDISMODULE_TARGET
+#define rm_malloc RedisModule_Alloc
+#define rm_calloc RedisModule_Calloc
+#define rm_realloc RedisModule_Realloc
+#define rm_free RedisModule_Free
+#else
+#define rm_malloc malloc
+#define rm_calloc calloc
+#define rm_realloc realloc
+#define rm_free free
+#endif
+
 static inline void *defragPtr(RedisModuleDefragCtx *ctx, void *ptr) {
     void *tmp = RedisModule_DefragAlloc(ctx, ptr);
     return tmp ? tmp : ptr;
 }
+
+static inline int rm_vasprintf(char **restrict str, char const *restrict fmt, va_list args) {
+    va_list args_copy;
+    va_copy(args_copy, args);
+    int needed = vsnprintf(NULL, 0, fmt, args) + 1;
+    *str = rm_malloc(needed);
+    int res = vsprintf(*str, fmt, args_copy);
+    va_end(args_copy);
+    return res;
+}
+
+static inline int rm_asprintf(char **restrict str, char const *restrict fmt, ...) {
+    va_list args;
+    va_start(args, fmt);
+    int res = rm_vasprintf(str, fmt, args);
+    va_end(args);
+    return res;
+}
+
+static inline int SetCommandAcls(RedisModuleCtx *ctx, char const *cmd, char const *acls,
+                                 char const *module) {
+    RedisModuleCommand *command = RedisModule_GetCommand(ctx, cmd);
+    if (command == NULL) {
+        RedisModule_Log(ctx, "error", "Failed to get command %s", cmd);
+        return REDISMODULE_ERR;
+    }
+
+    bool acls_empty = strcmp(acls, "") == 0;
+    char const *categories = acls_empty ? module : ({
+        char *c = NULL;
+        rm_asprintf(&c, "%s %s", acls, module);
+        c;
+    });
+
+    if (RedisModule_SetCommandACLCategories(command, categories) != REDISMODULE_OK) {
+        RedisModule_Log(ctx, "error", "Failed to set ACL categories %s for command %s", categories,
+                        cmd);
+        return REDISMODULE_ERR;
+    }
+    if (!acls_empty) {
+        rm_free((void *)categories);
+    }
+    return REDISMODULE_OK;
+}
+
+#define RegisterCommandWithModesAndAcls(ctx, cmd, f, mode, acls)                                   \
+    if (RedisModule_CreateCommand(ctx, cmd, f, mode, 1, 1, 1) != REDISMODULE_OK ||                 \
```

Comment:
add #error if MODULE_ACL_CATEGORY_NAME doesn't exists

Reply:
thinking more on this, i don't think we can #error here because the preprocessor doesn't expand the macro. i think your original idea of asserting here is necessary

Reply:
thinking even more on this, if the identifier is not defined, this will miscompile regardless. any assert on the token will not even be called because the line calling the assert will miscompile

Reply:
I didn't even understand which error @eranhd wanted exactly, so I am just leaving it as-is.

Reply:
Ah, I see now. Why don't we pass it here as a parameter? This will be lengthy - yes. But this will also make macro pure in the functional meaning, which doesn't depend on any environment and the outcome of which depends solely on the arguments. I see the code further constantly redefining a name and all this dependency on the environment... I don't like, but do as you wish, I guess. But then, the preprocessor should be able to see the string literal if we pass it as a string literal, shouldn't it? And the defined strings can be checked at a preprocessing time as Eran suggested.

Reply:
the suggesting was to `#error` if the macro were undefined, but that is given to us for free. we miscompile if the macro is undefined, which is literally the same thing.

Reply:
This is fine; at least, this is just one use-case for this macro, and only one "global" is used in this way.
About the `#error`, a clear message instead of being unable to compile something, I think, is a little better. Also, some code users may not understand what's wrong in the future, but if we explicitly tell them via the `#error` macro what is wrong, they will immediately know what to do.

Reply:
unfortunately it is impossible to conditionally apply an `#error` based on specific invocations of a macro. the `#error` is parsed by the preprocessor before the macro is ever invoked, so it will always be invoked because the symbol is not defined when the preprocessor parses the macro definition.